File: as1/h2o/h2o.py
from threading import Semaphore
from threading import Thread
import random
import time

class Barrier:
    def __init__(self, n):
        self.n = n
        self.count = 0
        self.mutex = Semaphore(1)
        self.turnstile = Semaphore(0)
        self.turnstile2 = Semaphore(0)
    def phase1(self):
        self.mutex.acquire()
        self.count += 1
        if self.count == self.n:
            for i in range(self.n):
                self.turnstile.release()
        self.mutex.release()
        self.turnstile.acquire()
    def phase2(self):
        self.mutex.acquire()
        self.count -= 1
        if self.count == 0:
            for i in range(self.n):
                self.turnstile2.release()
        self.mutex.release()
        self.turnstile2.acquire()
    def wait(self):
        self.phase1()
        self.phase2()

class Bonder:
    def __init__(self):
        self.bonded = 0
        self.currentOxy = 0
        self.currentHyd = 0

    def bond(self, identifier):
        if identifier == "oxygen":
            self.currentOxy += 1
        elif identifier == "hydrogen":
            self.currentHyd += 1
        else:
            raise ValueError("Invalid identifier")

        if self.currentOxy > 1 or self.currentHyd > 2:
            raise Exception("Invalid bond")
        elif self.currentOxy == 1 and self.currentHyd == 2:
            self.currentOxy = 0
            self.currentHyd = 0
            self.bonded += 1
            print("{:d} bonded.".format(self.bonded))


OXYGEN = 10_000
HYDROGEN = 2 * OXYGEN

# primitives
mutex = Semaphore(1)
oxygen = 0
hydrogen = 0
barrier = Barrier(3)
oxyQueue = Semaphore(0)
hydQueue = Semaphore(0)

bonder = Bonder()

def main():

    threads = []

    for i in range(OXYGEN):
        oxy_t = Thread(target=oxy, args=())
        threads.append(oxy_t)

    for i in range(HYDROGEN):
        hyd_t = Thread(target=hyd, args=())
        threads.append(hyd_t)

    random.shuffle(threads)

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

def oxy():
    global oxygen
    global hydrogen

    mutex.acquire()
    oxygen += 1
    if hydrogen >= 2:
        hydQueue.release()
        hydQueue.release()
        hydrogen -= 2
        oxyQueue.release()
        oxygen -= 1
    else:
        mutex.release()

    oxyQueue.acquire()
    bonder.bond("oxygen")

    barrier.wait()
    mutex.release()

def hyd():
    global oxygen
    global hydrogen

    mutex.acquire()
    hydrogen += 1
    if hydrogen >= 2 and oxygen >= 1:
        hydQueue.release()
        hydQueue.release()
        hydrogen -= 2
        oxyQueue.release()
        oxygen -= 1
    else:
        mutex.release()

    hydQueue.acquire()
    bonder.bond("hydrogen")

    barrier.wait()


if __name__ == "__main__":
    start_time = time.time()
    main()
    print("--- %s seconds ---" % (time.time() - start_time))

File: as1/h2o/h2o.go
package main

import (
    "fmt"
    "time"
    "sync"
    "math/rand"
)


type Oxygen struct {}
type Hydrogen struct {}

type Water struct {
    index int
    oxygen Oxygen
    hydrogen1 Hydrogen
    hyrdogen2 Hydrogen
}

// constants
const OXYGEN = 10000
const HYDROGEN = 2 * OXYGEN
const TOTAL = OXYGEN + HYDROGEN

// channels
var oxyQueue = make(chan Oxygen, OXYGEN)
var hydQueue = make(chan Hydrogen, HYDROGEN)

func main() {
    startTime := time.Now()

    rand.Seed(time.Now().UTC().UnixNano())

    wg := new(sync.WaitGroup)
    wg.Add(1)

    go bonder(wg)

    currentOxy := 0
    currentHyd := 0
    for i := 0; i < TOTAL; i++ {
        random := rand.Intn(2)
        if ((random == 0 && currentOxy != OXYGEN) || currentHyd == HYDROGEN) {
            oxyQueue <- Oxygen {}
            currentOxy ++
        } else {
            hydQueue <- Hydrogen {}
            currentHyd ++
        }
    }

    wg.Wait()

    fmt.Printf("--- %s ---\n", time.Since(startTime))
}

func bonder(wg *sync.WaitGroup) {
    defer wg.Done()

    MOLECULES := TOTAL / 3
    for i := 1; i < MOLECULES + 1; i += 1 {
        oxygen := <-oxyQueue
        hydrogen1 := <-hydQueue
        hydrogen2 := <-hydQueue

        water := Water {
            i,
            oxygen,
            hydrogen1,
            hydrogen2,
        }

        fmt.Printf("%d bonded.\n", water.index)
    }
}
